datafusion/core/src/physical_plan/joins/hash_join.rs

```diff
@@ -563,16 +563,24 @@ pub fn update_hash(
     // insert hashes to key of the hashmap
     for (row, hash_value) in hash_values.iter().enumerate() {
         let item = hash_map
-            .0
+            .map
             .get_mut(*hash_value, |(hash, _)| *hash_value == *hash);
-        if let Some((_, indices)) = item {
-            indices.push((row + offset) as u64);
+        if let Some((_, index)) = item {
+            // Already exists: add index to next array
+            let prev_index = *index;
+            // Store new value inside hashmap
+            *index = (row + offset + 1) as u64;
```

Comment:
Additions become O(1) by holding the end of the chain, right?

Reply:
I think the reason is that while iterating over the hashes/indices we get the latest index (which contains both the value **and** points to the previous index each time) as a constant time operation. Not sure how it would work when holding the chain start in the map as we have to iterate the map first to get to the last?

It would be possible (though seems not beneficial for the normal hash join) to also keep the start of the chain in the hashmap.

Reply:
Yeah, there is no gain for the usual hash join, but pruning becomes much more expensive if I do not have the beginning. I think I will not push for it, for now, let s have separate hashmap paradigms.

Reply:
> Additions become O(1) by holding the end of the chain, right?

Yes, this way `next[value - 1]` contains the previous value, and the next value / index can be found in the same way again.